### PR TITLE
feat: test saved tokens

### PR DESCRIPTION
**Pull request by gikf**
- [x] I have read [Chapter's contributing guidelines](https://github.com/freeCodeCamp/chapter/blob/main/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update README.md`).
- [x] My pull request targets the `main` branch of Chapter.

Ref #1885

---

- Adds ability to run test on saved (valid so far) tokens, via button in `dashboard/calendar`.
- So far test tries to obtain from api `calendarList`.
- ~~Draft mainly because there's currently awkwardness while ensuring all valid tokens are tested, and the right token would be marked as invalid, if it fails. And as usual I'm a bit confused with the async looping.~~

### CODE REVIEW

File: server/src/services/InitGoogle.ts [hunk 1]
```diff
@@ -159,14 +163,35 @@ async function createCredentialedClient() {
...
-  return calendar({ version: 'v3', auth });
+export async function createCalendarApi(
+  createClient: () => Promise<OAuth2Client> = async () =>
+    createCredentialedClient(
+      await prisma.google_tokens.findFirstOrThrow({
+        where: { is_valid: true },
+      }),
+    ),
+) {
+  return calendar({ version: 'v3', auth: await createClient() });
```
Comment (ojeytonwilliams) [inline server/src/services/InitGoogle.ts:174]:
```suggestion
export async function createCalendarApi(tokenInfo?: TokenInfo) {
  const tokens =
    tokenInfo ??
    (await prisma.google_tokens.findFirstOrThrow({
      where: { is_valid: true },
    }));
  const auth = await createCredentialedClient(tokens);
  return calendar({ version: 'v3', auth });
```
I found passing around `createClient` a little confusing and this approach seems more direct to me.

File: server/src/services/InitGoogle.ts [hunk 2]
```diff
@@ -159,14 +163,35 @@ async function createCredentialedClient() {
...
+  });
+  const apis = [];
+  for (const tokenInfo of tokensInfo) {
+    apis.push({
+      tokenId: tokenInfo.id,
+      createApi: () =>
+        createCalendarApi(() => createCredentialedClient(tokenInfo)),
+    });
+  }
+  return apis;
```
Comment (ojeytonwilliams) [inline server/src/services/InitGoogle.ts:189]:
```suggestion
  return await Promise.all(
    tokensInfo.map(async (tokenInfo) => ({
      tokenId: tokenInfo.id,
      calendarApi: await createCalendarApi(tokenInfo),
    })),
  );
```
I prefer this approach for a few reasons. It's functional, it flattens the Promises (it's a Promise, not a Promise of Promises) and finally you don't have to name as many things :D

File: server/src/services/Google.ts [hunk 3]
```diff
@@ -231,3 +236,21 @@ export async function deleteCalendarEvent({
     }),
   );
 }
+
+export async function testTokens() {
+  const apis = await createCalendarApis();
+
+  await Promise.all(
+    apis.map(async ({ tokenId, createApi }) => {
+      const calendarApi = await createApi();
```
Comment (ojeytonwilliams) [inline server/src/services/Google.ts:245]:
```suggestion
    apis.map(async ({ tokenId, calendarApi }) => {
```
The refactoring in `initGoogle` lets us write a simpler loop here.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: server/src/services/InitGoogle.ts (hunk 2 above)
Comment (gikf) [reply server/src/services/InitGoogle.ts]: :D I need to do this more often, because trying to find various names is not coming that great.

**Comment by gikf** [comment]
@ojeytonwilliams Thanks, that makes things clearer and simpler.

**Comment by ojeytonwilliams** [comment]
In case you're interested, I came across the 'trick' of avoiding naming things from Brian Will's youtube channel: https://www.youtube.com/@briantwill I forget which exactly which video, I'm afraid, but the channel as a whole is great.

Anyways, thanks for making those changes!

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by gikf** [comment]
Thanks, I think I found that portion in https://www.youtube.com/watch?v=2-G5WTghjYk _avoid naming as much as possible_. Other interesting insights as well :rocket:

**Pull request merged**